community/neo4j/src/test/java/ConcurrentChangesOnEntitiesTest.java

```diff
@@ -0,0 +1,249 @@
+
+        assertDatabaseConsistent();
+    }
+
+    private long initWithNode( GraphDatabaseService db )
+    {
+        try ( Transaction tx = db.beginTx() )
+        {
+            Node theNode = db.createNode();
+            long id = theNode.getId();
+            tx.success();
+            return id;
+        }
+
+    }
+
+    private long initWithRel( GraphDatabaseService db )
+    {
+        try ( Transaction tx = db.beginTx() )
+        {
+            Node node = db.createNode();
+            node.setProperty( "a", "prop" );
+            Relationship rel = node.createRelationshipTo( db.createNode(), DynamicRelationshipType.withName( "T" ) );
+            long id = rel.getId();
+            tx.success();
+            return id;
+        }
+    }
+
+    private Thread newThreadForNodeAction( final long nodeId, final Consumer<Node> nodeConsumer )
+    {
+        return new Thread( new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                try ( Transaction tx = db.beginTx() )
+                {
+                    Node node = db.getNodeById( nodeId );
+                    barrier.await();
+                    nodeConsumer.accept( node );
+                    tx.success();
+                }
+                catch ( Exception e )
+                {
+                    ex.set( e );
+                }
+            }
+        } );
+    }
+
+    private Thread newThreadForRelationshipAction( final long relationshipId, final Consumer<Relationship> relConsumer )
+    {
+        return new Thread( new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                try ( Transaction tx = db.beginTx() )
+                {
+                    Relationship relationship = db.getRelationshipById( relationshipId );
+                    barrier.await();
+                    relConsumer.accept( relationship );
+                    tx.success();
+                }
+                catch ( Exception e )
+                {
+                    ex.set( e );
+                }
+            }
+        } );
+    }
+
+    private void startAndWait( Thread t1, Thread t2 ) throws Exception
+    {
+        t1.start();
+        t2.start();
+
+        t1.join();
+        t2.join();
+
+        if ( ex.get() != null )
+        {
+            throw ex.get();
+        }
+    }
+
+    private void assertDatabaseConsistent() throws IOException
+    {
+        LogProvider logProvider = FormattedLogProvider.toOutputStream( System.out );
+        try
+        {
+            ConsistencyCheckService.Result result = new ConsistencyCheckService().runFullConsistencyCheck(
+                    testDirectory.graphDbDir(), new Config(), ProgressMonitorFactory.textual( System.err ),
+                    logProvider );
+            assertTrue( result.isSuccessful() );
+        }
+        catch ( ConsistencyCheckIncompleteException e )
+        {
+            throw new RuntimeException( e );
```

Comment:
It's better to fail test here with some short message, and printing out exception as well